### PR TITLE
support scan empty projection 

### PR DESCRIPTION
**Pull request by haohuaijin**
## Which issue does this PR close?

related to #3214 

## Rationale for this change

When I want to do the query like
```
select random() from test;
```
I encounter `Arrow error: External error: Arrow error: Invalid argument error: must either specify a row count or at least one column`.  I found that the error was occurring because, after reading a `RecordBatch`, we will to reproject the `RecordBatch` use the partition column, and it did not support an empty projection. see below.
https://github.com/apache/arrow-datafusion/blob/ba50a8b178eece7e79b100d0b73bdc9d6d3ec6d5/datafusion/core/src/datasource/physical_plan/file_scan_config.rs#L342

## What changes are included in this PR?

- change `file_scan_config.rs` and `cross_join.rs` to support empty `RecordBatch`
- change `push_down_projection` to support scan table with no projection

### CODE REVIEW

**Comment by haohuaijin** [comment]
very strange, `avro.slt` did not report an error when I ran `cargo test` locally.

**Comment by alamb** [comment]
> very strange, avro.slt did not report an error when I ran cargo test locally.


I think that is because avro is not a default feature

You probably need to run `cargo test --features=avro` to run it locally

**Review by alamb — APPROVED** [review]
Looks great to me. Thank you.

cc @Dandandan

File: datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs [hunk 1]
```diff
@@ -149,10 +149,11 @@ impl TableProvider for CustomProvider {
     async fn scan(
         &self,
         _state: &SessionState,
-        _: Option<&Vec<usize>>,
+        projection: Option<&Vec<usize>>,
         filters: &[Expr],
         _: Option<usize>,
     ) -> Result<Arc<dyn ExecutionPlan>> {
+        let projection = projection.unwrap();
```
Comment (Weijun-H) [inline datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs:156]:
```suggestion
        let projection = projection.unwrap_or_else(Vec::new);
```

Comment (Weijun-H) [reply datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs:156]: It will panic if projection is `None`

**Review by Weijun-H — CHANGES_REQUESTED** [review]

File: datafusion/optimizer/src/push_down_projection.rs [hunk 2]
```diff
@@ -149,10 +149,6 @@ impl OptimizerRule for PushDownProjection {
             {
                 let mut used_columns: HashSet<Column> = HashSet::new();
                 if projection_is_empty {
-                    let field = find_small_field(scan.projected_schema.fields()).ok_or(
```
Comment (Dandandan) [inline datafusion/optimizer/src/push_down_projection.rs:152]: I wonder if we can remove it as well for `LogicalPlan::Values`?

**Comment by Dandandan** [comment]
FYI @ch-sc

File: datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs [hunk 3]
```diff
@@ -182,7 +183,11 @@ impl TableProvider for CustomProvider {
                 };
 
                 Ok(Arc::new(CustomPlan {
-                    schema: self.zero_batch.schema(),
+                    // schema: self.zero_batch.schema(),
```
Comment (Dandandan) [inline datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs:186]:
```suggestion
```

File: datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs (hunk 1 above)
Comment (Dandandan) [reply datafusion/core/tests/custom_sources_cases/provider_filter_pushdown.rs:156]: `projection.unwrap_or_default()`

File: datafusion/physical-plan/src/joins/cross_join.rs [hunk 4]
```diff
@@ -347,13 +348,14 @@ fn build_batch(
         })
         .collect::<Result<Vec<_>>>()?;
 
-    RecordBatch::try_new(
+    RecordBatch::try_new_with_options(
```
Comment (Dandandan) [inline datafusion/physical-plan/src/joins/cross_join.rs:351]: I wonder if we have more operators with empty inputs that use `RecordBatch::try_new` that should move to use `RecordBatch::try_new_with_options`

File: datafusion/sqllogictest/test_files/subquery.slt [hunk 5]
```diff
@@ -695,7 +695,7 @@ logical_plan
 Projection: __scalar_sq_1.COUNT(*) AS b
 --SubqueryAlias: __scalar_sq_1
 ----Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1)) AS COUNT(*)]]
-------TableScan: t1 projection=[t1_id]
+------TableScan: t1 projection=[]
```
Comment (Dandandan) [inline datafusion/sqllogictest/test_files/subquery.slt:698]: 🎉

**Review by Dandandan — APPROVED** [review]

File: datafusion/optimizer/src/push_down_projection.rs (hunk 2 above)
Comment (haohuaijin) [reply datafusion/optimizer/src/push_down_projection.rs:152]: yes, we can remove it.

File: datafusion/physical-plan/src/joins/cross_join.rs (hunk 4 above)
Comment (haohuaijin) [reply datafusion/physical-plan/src/joins/cross_join.rs:351]: I have thought about this too, but I didn't come up with any more tests😥.

Comment (Dandandan) [reply datafusion/physical-plan/src/joins/cross_join.rs:351]: I think it is fine to merge as is and see if we get any regressions (hopefully not), those should be easy to fix in a similar manner.

**Comment by Dandandan** [comment]
Thank you @haohuaijin !

**Pull request merged**